nidmviewerfsl/viewer.py

```diff
@@ -372,7 +381,215 @@ def statisticImageString(statImage):
 	
 		return("Z (Gaussianised T/F)")
 
-	
+def formatClusterStats(g, excName):
+
+        
+        clusterData = {}
+        
+        #---------------------------------------------------------------------------------------------------------
+        #First we gather data for peaks table.
+        #---------------------------------------------------------------------------------------------------------
+
+        #We must include cluster p values in the query.
+        peak_query = """prefix nidm_SupraThresholdCluster: <http://purl.org/nidash/nidm#NIDM_0000070>
+                       prefix nidm_clusterSizeInVoxels: <http://purl.org/nidash/nidm#NIDM_0000084>
+                       prefix nidm_clusterLabelID: <http://purl.org/nidash/nidm#NIDM_0000082>
+                       prefix nidm_equivalentZStatistic: <http://purl.org/nidash/nidm#NIDM_0000092>
+                       prefix prov: <http://www.w3.org/ns/prov#>
+                       prefix nidm_coordinateVector: <http://purl.org/nidash/nidm#NIDM_0000086>
+                       prefix nidm_pValueUncorrected: <http://purl.org/nidash/nidm#NIDM_0000116> 
+                       prefix nidm_pValueFWER: <http://purl.org/nidash/nidm#NIDM_0000115> 
+                       prefix nidm_qValueFDR: <http://purl.org/nidash/nidm#NIDM_0000119>
+                       
+                       
+                       SELECT ?peakStat ?clus_index ?loc
+
+                       WHERE {{?exc a nidm_ExcursionSetMap: . ?clus prov:wasDerivedFrom ?exc . ?clus a nidm_SupraThresholdCluster: .
+                               ?exc prov:atLocation ?conMap . ?clus nidm_clusterLabelID: ?clus_index .
+                               ?peak prov:wasDerivedFrom ?clus . ?peak nidm_equivalentZStatistic: ?peakStat .
+                               ?peak prov:atLocation ?locObj . ?locObj nidm_coordinateVector: ?loc .}
+
+                       FILTER(STR(?conMap) = '""" + excName + """'^^xsd:string)}"""
+                
+        #Run the peak query
+        peakQueryResult = g.query(peak_query)
+
+        #Retrieve query results.
```

Comment:
I am unsure what you mean. The list returned by a query isn't stable, running a query once does not give the same ordered list as running it a second time, (unless you perform some sort of sort using SPARQL's 'ORDER BY'). 

To work out which peak locations are associated to which p-values and to work out which cluster they belong to this query retrieves all of these features all at once (as although the ordering of P-Values might change each time you run the query it's still the same relative to the ordering of the coordinates). 

These lines run the query using the rdflib engine and then separate the results of the query out into a list of P-Values, peak locations and cluster indices. 

I could have done 3 queries seperately and used SPARQL's 'ORDER BY' option to make sure the query results all have the same relative ordering but that would involve 3 separate sort permutations being calculated and I thought this instead would be faster.

I hope this answers your question. Please let me know if you think the way I have implemented this could be improved upon!